language/vm/src/access.rs

```diff
@@ -156,11 +156,19 @@ pub trait ModuleAccess: Sync {
         &self,
         field_count: MemberCount,
         first_field: FieldDefinitionIndex,
-    ) -> &[FieldDefinition] {
+    ) -> Result<&[FieldDefinition], VMStaticViolation> {
         let first_field = first_field.0 as usize;
         let field_count = field_count as usize;
-        let last_field = first_field + field_count;
-        &self.as_module().as_inner().field_defs[first_field..last_field]
+        if let Some(last_field) = first_field.checked_add(field_count) {
```

Comment:
I am not sure I follow the intent here so let me comment, help me understand the intent and let me know your thoughts.
What we are doing here is to check that the addition does not overflow a usize, however the inputs are 2 u16 (if I remember correctly) so that seems excessive.
Also this code would panic anyway if the indices were out of bounds.
I am wondering if the intent was to make sure the slice "exists".
If the code was something along the lines of
```
match self.as_module().as_inner().field_defs.get(first_field..first_field + field_count) {
    Some(fields) => Ok(fields),
    None => Err(...)
}
```
we would have protected from any panic this function may generate.
However I think this code was supposed to be called after structural checks of the binary so that condition should not be possible.
I don't know if there is some `assume` we could use and I can imagine it's no easy task.
Also this change still panics in a lot of scenarios (all the `expect`) even though I appreciate the most critical code path (that of the "loader") is protected and returns an `Err`.

Overall though my feeling is that if we understood how to handle a panic and collect useful diagnostic/debug info this would be a great place to let the panic go.
I wonder what MIRAI position is on the subject

Reply:
This seems like a good example of a case where one would use an assume. I.e. there is a prior check that ensures this cannot happen, but the nature of the prior check is such that the verifier cannot know about  it without the assume. And, as you observe, there is no need to (double) check the assumption at runtime, at this point, since there are later runtime checks that would fail, so there is insurance against later changes that may violate this assumption.

At best, doing the check here would make it easier to debug things.

Reply:
Thanks for the insight @dariorussi and @hermanventer. As Herman suggested, I'm going to switch this check to an assumption.